File: C++ Code/Data Structure/Binary Tree/Pre-Order Traversal.cpp
// WAP to build a binary tree and print data in Pre-Order Traversal using recursion.
#include<iostream>
#include<vector>
using namespace std;
class node{
  public:
    int data;
    node* left;
    node* right;
    
  node(int d){
    this->data=d;
    this->left=NULL;
    this->right=NULL;
  }  
};
node* build(node* root){
  int data;
  cout<<"Enter Data : ";
  cin>>data;
  if(data==-1)
    return root;
  root=new node(data);
  cout<<"For left of "<<data<<" , ";
  root->left=build(root->left);
  cout<<"For right of "<<data<<" , ";
  root->right=build(root->right);
  return root; 
}
void POT(vector<int> &arr, node* root){
  if(!root)
    return ;
  arr.push_back(root->data);
  POT(arr,root->left);  
  POT(arr,root->right);
}
int main(){
  cout<<"NOTE : Press -1 for leaf node.\n\n";
  node* root=NULL;
  root=build(root);
  vector<int>arr;
  POT(arr,root);
  cout<<"\n\nPre-Order Traversal ;\n";
  for(int i:arr)
    cout<<i<<" ";
  return 0;
}

File: Javascript Code/Binary Tree/Preorder Traversal.js
/*
  Given the root of a binary tree, return the preorder traversal of its nodes' values.
  
  Constraints:
  -> The number of nodes in the tree is in the range [0, 100].
  -> -100 <= Node.val <= 100
*/
var preorderTraversal = function(root) {
  let ans=[];
  solve(root,ans);
  return ans;
};
function solve(root,ans){
  if(!root)
    return;
  ans.push(root.val);
  solve(root.left,ans);
  solve(root.right,ans);
}